mcs/class/corlib/System.Threading/Interlocked.cs

```diff
@@ -50,7 +50,27 @@ public static partial class Interlocked
 
 		[ReliabilityContractAttribute (Consistency.WillNotCorruptState, Cer.Success)]
 		[MethodImplAttribute(MethodImplOptions.InternalCall)]
-		public extern static object CompareExchange(ref object location1, object value, object comparand);
+		extern static void CompareExchange_object_icall (ref object location1, ref object value, ref object comparand, ref object result);
+
+		[ReliabilityContractAttribute (Consistency.WillNotCorruptState, Cer.Success)]
+		public static object CompareExchange (ref object location1, object value, object comparand)
+		{
+			// This avoids coop handles, esp. on the output which would be particularly inefficient.
+			// Passing everything by ref is equivalent to coop handles -- ref to locals at least.
+			//
+			// location1's treatment is unclear. But note that passing it by handle would be incorrect,
+			// as it would use a local alias, which the coop marshaling does, to avoid the unclarity here,
+			// that of a ref being to a managed frame vs. a native frame. Perhaps that could be revisited.
+			//
+			// So there a hole here, that of calling this function with location1 being in a native frame.
+			// Usually it will be to a field, static or not, and not even to managed stack.
+			//
+			// This is usually intrinsified. Ideally it is always intrinisified.
+			//
+			object result;
```

Comment:
This probably has to be initialized, or `out`.

Reply:
I know it is less efficient, but I went with initialized.
I get nervous about `out` parameters, esp. with all the wrappers. `ref` is very safe.